add onlyIfDevelopment entry file for easier webpack config

agilgur5 opened the pull request:
- add documentation for this option and leave the previous installation
  docs as an alternative "custom installation"

I based the docs off of the [previous commit that mentioned entry files](https://github.com/G-Rath/webpack-serve-overlay/commit/a99448cab7dc2facdae3fc948a17d995fee16f1a), with a few modifications:
- Added a link to webpack `entry` config since it's a bit of an advanced topic
- Removed `require.resolve` as I don't believe it's necessary (I didn't need to use it in [my webpack config](https://github.com/agilgur5/front-end-base/tree/only-if-production), though I'm not sure if there's a circumstance in which it would be necessary)

Sorry, my formatter removed whitespace at the end of lines automatically, I can remove those from the commit if needed.

Having looked at the past docs commit, I can see that my usage is a bit different than the way you've used it / seen it. I didn't add a new entry (and therefore no new `<script>` tag), I added it to the array in my entry, i.e.

```
entry: {
  main: ['webpack-serve-overlay/onlyIfProduction', './main.js']
}
```
See the link above for full context in my boilerplate repo.
**EDIT**: I definitely think having it as a separate / new entry is the ideal usage as separate entrypoints / script tags ensures the overlay always shows up as its separate, isolated bundle will never fail to compile. In that case, excluding the overlay script tag from production HTML would also be ideal to have as little production impact as possible. Even with `onlyIfProduction`, webpack's bundling code will still create a tiny bundle of JS that does nothing (not optimal, but not particularly harmful either).

Related to #5

agilgur5 committed:
add onlyIfDevelopment entry file for easier webpack config

- add documentation for this option and leave the previous installation
  docs as an alternative "custom installation"

G-Rath commented on README.md:
````diff
@@ -13,7 +13,15 @@ Install the package:
+Then add the following entry to the [`entry` object](https://webpack.js.org/configuration/entry-context/#entry) in your webpack config:
+
+```javascript
+overlay: 'webpack-serve-overlay/onlyIfProduction'
````
I think this file should be called `onlyIfDevelopment`, since the idea is that it's only included if `NODE_ENV === 'development'`, not 'production'.

To me it currently gives the impression that it'll only do something when in production 😉

agilgur5 replied:
Ah good point. I think I just went with the name you suggested and didn't even think about that haha

G-Rath replied:
Ya, I suggested that name b/c I thought the check was `=== 'production`, but it's the other way around XD

G-Rath commented on README.md:
```diff
@@ -23,11 +31,11 @@ if(process.env.NODE_ENV === 'development') {
 
 ## Configuration
 
-The overlay works by using a WebSocket that connects to `webpack-serve` à la `webpack-hot-client`.  
```
The double space here was deliberate, as it's an indicator in markdown that the following line should be on a newline. Removing it causes the next line to be rendered on the same line as this line.

(I'm guessing your formatter isn't markdown-aware?)

agilgur5 replied:
Oh right, I forget that's a really unintuitive feature of Markdown (near invisible trailing whitespace used for actual formatting); I tend to use `<br>` tags instead. 

(I guess not :/ will have to see if there's a better plugin for VSCode)

G-Rath reviewed (changes requested):
Looks pretty good.

Sorry it took a couple of days for me to get around to looking this over!

I'll give it a quick test with using `entry`, but aside from the two requested changes, it's looking pretty good!

G-Rath commented:
> Removed require.resolve as I don't believe it's necessary

I think this was based off the use of `path.resolve` with a lot of the other path-related things in webpack.

Having not had much experience with `entry`, I can't say if it's needed or not - it could very well have been the whole reason I was having problems.

I'm happy to defer to you for advise on the best way to handle the entry side of things 😃

G-Rath commented on README.md:
````diff
@@ -13,7 +13,15 @@ Install the package:
 ```
 
-Then require the overlay at the top of your `index.jsx` (or equivalent):
+Then add the following entry to the [`entry` object](https://webpack.js.org/configuration/entry-context/#entry) in your webpack config:
````
I've just tested doing this, and it seems to stop hot-reloading: 

```
    entry: {
        overlay: 'webpack-serve-overlay',
        bundle: path.resolve('src/index.jsx')
    },
```

Results in the following console.logs:

```
｢hot｣ webpack: Compiling (<unnamed compiler>)
10:15:42.412 log.js?60b4:59 ｢hot｣ App Updated, Reloading Modules
10:15:42.412 log.js?60b4:59 ｢hot｣ Checking for updates to the bundle.
10:15:42.488 log.js?60b4:59 ｢hot｣ No Modules Updated.
10:15:42.489 log.js?60b4:59 ｢hot｣ App is up to date.
```

If I remove the entry, and then make the same change, the page gets reloaded.

Could you check this out, and see if you can reproduce this on your end?

agilgur5 replied:
So I definitely can't reproduce that error with my config and [my boilerplate](https://github.com/agilgur5/front-end-base/) is now set up [similarly](https://github.com/agilgur5/front-end-base/blob/master/webpack.config.js) (slightly different from the [branch](https://github.com/agilgur5/front-end-base/blob/only-if-production/webpack.config.js) linked above). You can try running that and changing some code to see if it's potentially related to your environment.

Logs ok
```
log.js:59 ｢hot｣ Hot Module Replacement Enabled. Waiting for signal.
log.js:59 ｢hot｣ WebSocket connected
log.js:59 ｢hot｣ App updated. Recompiling components/app.js
log.js:59 ｢hot｣ webpack: Compiling (<unnamed compiler>)
log.js:59 ｢hot｣ App Updated, Reloading Modules
log.js:59 ｢hot｣ Checking for updates to the bundle.
log.js:59 ｢hot｣ No Modules Updated.
log.js:59 ｢hot｣ App is up to date.
````

Kinda hard to tell without more context (e.g. the rest of the config) what might be causing that issue for you. Wasn't totally clear from your comment, but just to make sure webpack can't currently detect changes in its own config

G-Rath replied:
That's what I expected; I'll provide more context soon, when I have some free time on my hands.

And don't worry, I know that you have to restart webpack for changes to take effect ;) 

Heres a gist containing how my webpack configs are setup:

https://gist.github.com/G-Rath/b441cd818940ff989edb216513b8d4f3

For the record, all of my configuration stuff was/is based off the examples given in the Webpack documentation, which is somewhat concerning if it's not compatible.

(Random fun fact: spread works with booleans, so you can just do `...(isProd || {errorOverlay: 'webpack-serve-overlay'})` instead of using `:?`)

agilgur5 replied:
Huh ok. There are like a 100 different ways of doing the same thing in webpack and before 2.0 the docs weren't the best place to find info but that's surprising nowadays ¯\\\_(ツ)_/¯ . 
Your config is a bit more complex than I expected, so not sure staring for errors would work well with it haha. If you have a `package.json` and a sample js file that replicates the problem (like the one in my boilerplate) I can test and see if it's reproduced in my environment, otherwise in your case I'd try to remove a bunch of the config and see if it still occurs.
Were you saying that HMR works with `require.resolve` but not the plain package name?? Because the gist isn't using `onlyIfDevelopment` so that's unrelated

(Indeed, another fun fact is that syntax causes problems on certain platforms like React Native because they have some "optimizations" that require you to have "a value" inserted there... I didn't investigate further into it but it was quite disorienting so I stuck to the more explicit syntax instead and kept using it everywhere)

agilgur5 replied:
@G-Rath so I started getting the problem of not hot-reloading as well, and found out that it's because ["`webpack-hot-client` automatically configur[es] the client script for entries on the _first_ entry only"](https://github.com/webpack-contrib/webpack-serve/issues/119#issuecomment-401502247).

I do think it's a significant enough and common enough problem (particularly based on all the confusion in that issue) that it's worth adding a note to the docs about, so I'll amend my PR with that.

agilgur5 commented:
Committed a new revision that should change it to `onlyIfDevelopment` and remove the formatter changes

agilgur5 committed:
(docs): add a note on how the entry should not be the first

- as doing so will break hot reloading for the default settings
- add link to the relevant webpack-serve issue & comment

agilgur5 commented:
Added a new commit with an addition to the docs regarding the hot reloading problem that occurs when this is the first entry

agilgur5 committed:
(docs): be more explicit with an entry obj with ellipsis

- should make the hot reload problem harder to stumble upon by making
  the intended entry config more explicit
